.changeset/fifty-bobcats-jog.md

```diff
@@ -0,0 +1,34 @@
+---
+'@graphql-tools/executor': major
+'@graphql-tools/utils': minor
+---
+
+Upgrade to non-duplicating Incremental Delivery format
+
+## Description
+
+GraphQL Incremental Delivery is moving to a [new response format without duplication](https://github.com/graphql/defer-stream-wg/discussions/69).
+
+This PR updates the executor within graphql-tools to follow the new format, a BREAKING CHANGE.
+
+Incremental Delivery has now been disabled for subscriptions, also a BREAKING CHANGE. The GraphQL Working Group has decided to disable incremental delivery support for subscriptions (1) to gather more information about use cases and (2) explore how to interleaving the incremental response streams generated from different source events into one overall subscription response stream.
+
+Library users can explicitly opt in to the older format by call `execute` with the following granular options:
+
+```ts
+const result = await execute({
+  ...,
+  deduplicateDefers: false,
+  sendIncrementalErrorsAsNull: true,
+  sendPathAndLabelOnIncremental: true,
+  errorWithIncrementalSubscription: false,
```

Comment:
When providing these parameters, I still receive `{"data":{},"pending":[{"id":"0","path":[]}],"hasNext":true}` instead of `{"data":{},"hasNext":true}`.

Seems like `pending` should also be omitted to be backward-compatible? (See failing tests in https://github.com/dotansimha/graphql-yoga/pull/3402)

_____

Regarding all the granular options - do you think we need to allow such granularly for the user? I think from our perspective they would only care about "old" spec and "new" spec, not on using only parts of the new spec? 🤔

So maybe instead we should use a version number? Do you see benefit in setting some of the flags and omitting others from an end-user and feedback-user perspective?

Reply:
I don’t think extra graphql fields are technically backwards incompatible as anything custom goes into extensions.

But I’m certainly open to an additional option to suppress the pending/completed which would require setting sendPathAndLabelOnIncremental to true.

We could also add a version option that implies a different combination of defaults so that the granular options could be simplified

I think I am also noticing that the errors are now sent after data, which is not preferred according to the specification it should also be fixed

Reply:
> I don’t think extra graphql fields are technically backward incompatible as anything custom goes into extensions.

It is still a different output than expected, and strict clients might react funky... I am unaware of any clients that do so right now, but I can already imagine this question coming up from a user. Since it seems straightforward to implement, let's add it.

______

> We could also add a version option that implies a different combination of defaults so that the granular options could be simplified

Is there some reference version name e.g. date or similar used within the working group context to differentiate the different drafts?

______

What do you think about grouping the defer/stream-related options within an object? For the user-facing API, they would be more organized and separated from the other options.

```typescript
execute({
  ...otherArgs,
  incrementalDelivery: {
    deduplicateDefers: false,
    sendIncrementalErrorsAsNull: true,
    sendPathAndLabelOnIncremental: true,
    errorWithIncrementalSubscription: false,
  }
})
```

In addition or instead, if the granular configuration of `deduplicateDefers`, `sendIncrementalErrorsAsNull`, `sendPathAndLabelOnIncremental`, `errorWithIncrementalSubscription` (and also an option for [dropping the additional new properties](https://github.com/ardatan/graphql-tools/pull/6243#discussion_r1703221495) - we can do something like the following:

```typescript
execute({
  ...otherArgs,
  incrementalDelivery:  "2024-08-working-draft"
})
```

Reply:
I reworked it as follows as described in the changeset:

Library users can explicitly opt in to the older format by call `execute` with the following option:

```ts
const result = await execute({
  ...,
  incrementalPreset: 'v17.0.0-alpha.2',
});
```

The default value for `incrementalPreset` when omitted is `'v17.0.0-alpha.3'`, which enables the new behaviors described above. The new behaviors can also be disabled granularly as follows:

```ts
const result = await execute({
  ...,
  deferWithoutDuplication: false,
  useIncrementalNotifications: false,
  errorOnSubscriptionWithIncrementalDelivery: false,
});
```

Reply:
There is some renaming and some consolidation => the use of ids is intimately connected to the use of pending/completed for notifications in whose absence it makes sense to send incremental errors as null entries. This seems more concise.